### PR TITLE
MAT-5808: Can edit fix

### PR DESCRIPTION
**Pull request by mcmcphillips**
## MADiE PR

Jira Ticket: [MAT-5808](https://jira.cms.gov/browse/MAT-5808)
(Optional) Related Tickets:

### Summary
This PR updates the library service to deliver librarySet objects with cql libraries to allow for checkUserCanEdit

### All Submissions
* [x] This PR has the JIRA linked.
* [x] Required tests are included and Code coverage is verified.
* [x] No extemporaneous files are included (i.e Complied files or testing results).
* [x] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [x] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [ ] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepository.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package gov.cms.madie.cqllibraryservice.repositories;
+
+import gov.cms.madie.models.common.Version;
+import gov.cms.madie.models.library.CqlLibrary;
+
+import java.util.List;
+
+public interface LibraryAclRepository {
+  /**
+   * Library is considered to be my measure if provided user is the owner of this measure or is
```
Comment (adongare) [inline src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepository.java]: The description needs to be updated

File: src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        .from("librarySet")
+        .localField("librarySetId")
+        .foreignField("librarySetId")
+        .as("librarySet");
+  }
+
+  @Override
+  public List<CqlLibrary> findAllMyLibraries(String userId) {
+
+    Criteria libraryCriteria = Criteria.where("createdBy").is(userId);
```
Comment (adongare) [inline src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java]: I do not think `createdBy` matters anymore. it should look for ownership of shared ACL

Comment (mcmcphillips) [reply src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java]: acls don't live on the library though. they're on libraryset.
@adongare, how would we decide library criteria if that's the case?

File: src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    List<CqlLibrary> results =
+        mongoTemplate
+            .aggregate(libraryAggregation, CqlLibrary.class, CqlLibrary.class)
+            .getMappedResults();
+    return results;
+  }
+
+  @Override
+  public List<CqlLibrary> findAllByCqlLibraryNameAndDraftAndVersion(
```
Comment (adongare) [inline src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java]: do not think this method is being used

File: src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    Aggregation librarySetAggregation = newAggregation(getLookupOperation(), matchOperation);
+    List<CqlLibrary> results =
+        mongoTemplate
+            .aggregate(librarySetAggregation, CqlLibrary.class, CqlLibrary.class)
+            .getMappedResults();
+    return results;
+  }
+
+  @Override
+  public List<CqlLibrary> findAllByCqlLibraryNameAndDraftAndVersionAndModel(
```
Comment (adongare) [inline src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java]: and this one either.

File: src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java (hunk 2 above)
Comment (chubert-sb) [reply src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java]: since you are already doing a join on librarySet it shouldn't matter. it'll still only pull the libraries with that owner or acl set

File: src/main/java/gov/cms/madie/cqllibraryservice/controllers/CqlLibraryController.java [hunk 5]
```diff
@@ -46,15 +50,27 @@ public ResponseEntity<List<CqlLibrary>> getCqlLibraries(
...
         filterByCurrentUser
-            ? cqlLibraryRepository.findAllByCreatedBy(username)
+            ? cqlLibraryRepository.findAllMyLibraries(username)
             : cqlLibraryRepository.findAll();
+    cqlLibraries.forEach(
+        l -> {
+          LibrarySet librarySet =
+              librarySetRepository.findByLibrarySetId(l.getLibrarySetId()).orElse(null);
+          l.setLibrarySet(librarySet);
+        });
```
Comment (chubert-sb) [inline src/main/java/gov/cms/madie/cqllibraryservice/controllers/CqlLibraryController.java:59]: I know there isn't a way around this right now because we have the transient tag on the LIbrarySet, but we should really look into fixing this in the future

File: src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    Criteria librarySetCriteria =
+        new Criteria()
+            .orOperator(
+                Criteria.where("librarySet.owner").is(userId),
+                Criteria.where("librarySet.acls.userId")
+                    .is(userId)
+                    .and("librarySet.acls.roles")
+                    .in(RoleEnum.SHARED_WITH));
+    MatchOperation matchOperation =
+        match(new Criteria().andOperator(libraryCriteria, librarySetCriteria));
```
Comment (chubert-sb) [inline src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java]: Oh actually since it's using the and operator with the createdBy this will actually cause a real problem.

If I transfer ownership to someone they will never see it because this will never return that library.

**Review by chubert-sb — DISMISSED** [review]

File: src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java (hunk 2 above)
Comment (mcmcphillips) [reply src/main/java/gov/cms/madie/cqllibraryservice/repositories/LibraryAclRepositoryImpl.java]: Conversation should be dated. createdby no longer referenced in impl.

**Review by RohitKandimalla — DISMISSED** [review]

File: src/main/java/gov/cms/madie/cqllibraryservice/services/VersionService.java [hunk 7]
```diff
@@ -89,9 +89,9 @@ private String libraryContentTemplate(String cqlLibraryName, Version version) {
   }
 
   private void validateCqlLibrary(CqlLibrary cqlLibrary, String username) {
-    if (!username.equalsIgnoreCase(cqlLibrary.getCreatedBy())) {
+    if (!username.equalsIgnoreCase(cqlLibrary.getLibrarySet().getOwner())) {
```
Comment (adongare) [inline src/main/java/gov/cms/madie/cqllibraryservice/services/VersionService.java]: I see the problem here if the user is  SHARED_WITH with user.

**Review by adongare — APPROVED** [review]
